xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -264,83 +158,68 @@ public Subchannel createSubchannel(CreateSubchannelArgs createSubchannelArgs) {
+    private final ChannelSecurityLbHelper lbHelper = new ChannelSecurityLbHelper();
     @Nullable
-    LoadBalancer edsBalancer;
-
-    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
-      this.helper = new EdsLoadBalancingHelper(helper,
-          new AtomicReference<SslContextProvider>());
-      this.resolvedAddresses = resolvedAddresses;
-    }
+    private LoadBalancer lb;
 
     @Override
-    public void onClusterChanged(ClusterUpdate newUpdate) {
+    public void onClusterChanged(ClusterUpdate update) {
       if (logger.isLoggable(XdsLogLevel.INFO)) {
-        logger.log(
-            XdsLogLevel.INFO,
-            "Received cluster update from xDS client {0}: "
+        logger.log(XdsLogLevel.INFO, "Received cluster update from xDS client {0}: "
                 + "cluster_name={1}, eds_service_name={2}, lb_policy={3}, report_load={4}",
-            xdsClient, newUpdate.getClusterName(), newUpdate.getEdsServiceName(),
-            newUpdate.getLbPolicy(), newUpdate.getLrsServerName() != null);
+            xdsClient, update.getClusterName(), update.getEdsServiceName(),
+            update.getLbPolicy(), update.getLrsServerName() != null);
       }
-      checkArgument(
-          newUpdate.getLbPolicy().equals("round_robin"), "can only support round_robin policy");
-
-      LoadBalancerProvider lbProvider = lbRegistry.getProvider(newUpdate.getLbPolicy());
-      Object lbConfig =
-          lbProvider.parseLoadBalancingPolicyConfig(ImmutableMap.<String, Object>of()).getConfig();
+      LoadBalancerProvider lbProvider = lbRegistry.getProvider(update.getLbPolicy());
       final EdsConfig edsConfig =
-          new EdsConfig(
-              /* clusterName = */ newUpdate.getClusterName(),
-              /* edsServiceName = */ newUpdate.getEdsServiceName(),
-              /* lrsServerName = */ newUpdate.getLrsServerName(),
-              new PolicySelection(lbProvider, ImmutableMap.<String, Object>of(), lbConfig));
-      if (isXdsSecurityEnabled()) {
-        updateSslContextProvider(newUpdate.getUpstreamTlsContext());
+          new EdsConfig(update.getClusterName(), update.getEdsServiceName(),
+              update.getLrsServerName(), new PolicySelection(lbProvider, null, null));
+      if (enableSecurity) {
+        updateSslContextProvider(update.getUpstreamTlsContext());
       }
-      if (edsBalancer == null) {
-        edsBalancer = lbRegistry.getProvider(EDS_POLICY_NAME).newLoadBalancer(helper);
+      if (lb == null) {
+        lb = lbRegistry.getProvider(XdsLbPolicies.EDS_POLICY_NAME).newLoadBalancer(lbHelper);
       }
-      edsBalancer.handleResolvedAddresses(
-          resolvedAddresses.toBuilder().setLoadBalancingPolicyConfig(edsConfig).build());
+      lb.handleResolvedAddresses(
+          ResolvedAddresses.newBuilder()
+              .setAddresses(Collections.<EquivalentAddressGroup>emptyList())
+              .setAttributes(
+                  Attributes.newBuilder()
```

Comment:
Why making this change? Should never remove any unrecognized attributes.

Reply:
The xDS LB tree implementation is internal, so we know what attributes should be there. Keeping unused/unexpected blobs around just prevent them from being GCed.

Reply:
No. (1) Child LBs in the xDS LB tree are plugins, and in theory any LB can be replaced another LB/LB tree, e.g. `RoundRobinLoadBalancer` can be replaced by `HealthCheckingRoundRobinLoadBalancer`. (2) You don't know what child policies will be supported in the future, so it will be hard to maintain the whole LB tree if a new child policy need another attributes. (3) We should follow the same rule for hierarchical LB tree in general.

GC is not a concern, they will be GCed when the whole LB tree is not referencing the attributes.

Reply:
> (1) Child LBs in the xDS LB tree are plugins, and in theory any LB can be replaced another LB/LB tree, e.g. RoundRobinLoadBalancer can be replaced by HealthCheckingRoundRobinLoadBalancer. (2) You don't know what child policies will be supported in the future, so it will be hard to maintain the whole LB tree if a new child policy need another attributes.

Why do you want to add code and implement something undefined yet? Currently the only useful attribute is the XdsClientPool. What's the different between this code and pass down the `ResolvedAddress` blob? If more attributes are added from upstream and are used by downstreams, just implement that logic when needed. Why do you want to implement a placeholder for something that is currently undefined?

> (3) We should follow the same rule for hierarchical LB tree in general.

A rule? What's that rule? What's the point of that rule?

Reply:
If the name resolver sends an attribute that the top-level LB does not recognize, it's supposed to be used for its grand children or transports. The name resolver should know what attributes it's adding. The name resolver should never add an attribute that is not used downstream. If the top-level LB does not recognize that attribute, it should still pass down to its children. Only when the top-level LB recognizes an attribute, and knows it is not needed for its children or grand children or transports, then it should not pass the attribute downstream.

One LB should never add a new attribute that is not used downstream. If one LB adds an attribute that its direct child does not recognize, it's supposed to be used for its grand children or transport. The LB should know what attributes it's adding. The LB should never add an attribute that is not used downstream. Only when the LB recognizes an attribute, and knows it is not needed for its children or grand children or transports, then it should not pass the attribute downstream.

Every layer knows what new attributes to add and knows what known attributes to be removed and should not remove attributes it does not recognize. So there is no chance unused attributes passed downstream.

Reply:
> If the name resolver sends an attribute that the top-level LB does not recognize, it's supposed to be used for its grand children or transports. The name resolver should know what attributes it's adding. The name resolver should never add an attribute that is not used downstream. If the top-level LB does not recognize that attribute, it should still pass down to its children. Only when the top-level LB recognizes an attribute, and knows it is not needed for its children or grand children or transports, then it should not pass the attribute downstream.

What is such an attribute here? Obviously no such a thing.

> One LB should never add a new attribute that is not used downstream. If one LB adds an attribute that its direct child does not recognize, it's supposed to be used for its grand children or transport. The LB should know what attributes it's adding. The LB should never add an attribute that is not used downstream. Only when the LB recognizes an attribute, and knows it is not needed for its children or grand children or transports, then it should not pass the attribute downstream.

I didn't add any new attribute here. CDS LB policy knows downstream needs the XdsClient, so it put that into the attribute. It doesn't put anything extra.

> Every layer knows what new attributes to add and knows what known attributes to be removed and should not remove attributes it does not recognize. So there is no chance unused attributes passed downstream.

Again, what is the unrecognized attribute here? Not any right? What trying to worry about something that does not exist?

Reply:
What attribute is not GCed? Not any right? Why worry about something that does not exist?

Reply:
> What attribute is not GCed?

Yeah, not any. So I am not saying that anymore.

Reply:
May let's get a second opinion. + @ejona86

Reply:
I reordered the quotes a bit to fix things together a bit more.

> The xDS LB tree implementation is internal, so we know what attributes should be there.

That's a very brittle view of a system. With that approach it is fine for _any_ change to require auditing all the code (which means _every_ change, since you don't when your change would impact other parts unless you audit).

I agree we could live with such things, but they are technical debt and should be strongly discouraged for our sanity and velocity. (Note, on a small scale (e.g., a method) this is fine. But the larger the "component" impacted the greater the cost and thus the more frequently it should be avoided.)

> The upstream just pass should just what's needed by downstream and nothing more.

That is a strongly incorrect argument. Upstream does not know what downstream needs.  We _know_ downstream will be pluggable in the future. Upstream does not have a full view of the system and to assert it _should_ is just silly. But also, this isn't really "upstream" this is "midstream" and we need upstream and downstream to be able to communicate without having midstream be aware of everything that's happening.

> This is much more elegant than blindly cache a blob and pass it to another component.

But that's the hierarchical design with the attributes. That's the _point_.

> Much more compact logic and avoid spreading unused data everywhere.

I don't see what logic would be necessary. Could you point out what logic is removed? It seems any logic relating to this should be trivial.

> This is much easier to understand and reason about.

Yes and no. It is easier to reason about _a single component_. It is harder to reason about the _system_. In both cases it may be harder to make changes in the future (it depends a lot on which types of changes whether it is a win or loss). Given that propagating the attributes here seems trivial for the component, it seems a very good tradeoff to benefit the system.

> Keeping unused/unexpected blobs around just prevent them from being GCed.

Would they actually be GCed? I'd generally expect expensive things to be retained by the higher levels in order to re-use them across updates.

Is there a specific object in particular or a reason to think this would be a problem? Optimizing for GC is optimizing and optimizing without a clear need is premature optimization. I could argue that we should optimize to reduce the allocation rate and so you should reuse the attributes. These attributes aren't all that heavily used, but where they are used it is important stuff and could significantly reduce overheads by sharing (e.g., health checks can be shared). So your microptimization here could have real-life negative consquences.

I don't think "optimization" is the correct argument. The change discussed here has correctness/semantics implications. Unless one option is _significantly_ worse, favoring correctness yields a much better system. So we really shouldn't bring up the GC argument at all unless there is a "clear and present danger" and there clearly is not such an issue here.

I don't think you should make subtle behavior changes during a refactor like this. Do the change before or after, or have a _really good_ reason why it has to be within the refactor. Refactors are risky and you should spend extra effort to avoid having behavior changes hidden within them. When a refactor has no behavior changes, it is _much_ easier to review as well.

> Why add code and implement something undefined yet?

This may be a valid argument. I don't know this code well enough to tell. It's a "how much work would it take to plumb the attributes" question. However, it was plumbed previously, so I would be doubtful that removing that plumbing makes sense.

> One LB should never add a new attribute that is not used downstream.

That's not 100% true. There are some attributes that are added "just in case." Health checking for example. But those attributes are needed by the LB adding them already, so there's no GC concern; the LB that added the attribute hangs on to that resource.

Consider the xdsClient. That's "big and expensive." But it doesn't help _anyone_ to strip it from the attributes, because the NameResolver is going to use the xdsClient even if no LB does. It is a _significant optimization_ to pass the xdsClient to child LB policies _just in case_ they need it. And if they don't need it, no harm was done; it's relatively free. (Yes, we know xdsClient is used a lot; but it could make sense to do the same thing even if it was a rare benefit. Now, we don't do that for all resources simply because that would be premature optimization and a waste of our development energy. So we do it just for "objects that make sense".)

Reply:
Thanks for your dedicated explanation! I am considering this more from the cosmetic perspective. Mostly I hate the pattern of the code that turns a data blob into builder, swap out some parts and then build to be used for the next components. The blob may carry excessive data that intended to be consumed only at this components to other components. Instead, I prefer the pattern of rebuilding the data blob passed to next components by adding in pieces taken from the current blob explicitly. Building the blob explicitly makes the code cleaners to understand and semantically less error-prone by avoid leaking excessive data to other components.

> > This is much more elegant than blindly cache a blob and pass it to another component.
> 
> But that's the hierarchical design with the attributes. That's the _point_.

Alright, what about we just plumb the attributes opaquely, but we would build the ResolvedAddresses by explicitly putting in the attributes. This seems to be a solution that is somewhat in the middle of what we are arguing.

Reply:
> Mostly I hate the pattern of the code that turns a data blob into builder, swap out some parts and then build to be used for the next components.

I'm sorry, but that's the design and pattern. Doing anything else is harmful.

> Instead, I prefer the pattern of rebuilding the data blob passed to next components by adding in pieces taken from the current blob explicitly.

I don't really know what to say but "too bad." I've discussed the problems with that approach and benefits to the add-to-the-blob approach. The design is already there and being used and to change it would mean a redesign; it's not just this one component your playing with, you're breaking the system it is apart of.

With the current approach the `Attribute.Key` may not even be publicly visible. So you _many not be able to copy it_ to a new object. That's a nice feature as we can use Java visibility to limit access to data.

The approach you discuss could be another design approach, but it is clearly not the current design for the system. You could argue for us to change the current design due to problems X, Y, Z, but you _can't_ change just this component without getting the worst of both approaches without any of the benefits. You should proceed with the current system's design and we only change code to a different "style" if we actually change the design of the system.

You can dislike the system, but that doesn't change "we have to work within the system." Yes, the system can be changed, but not like this.

> Building the blob explicitly makes the code cleaners to understand and semantically less error-prone by avoid leaking excessive data to other components.

This seems pure personal preference or speculation compared to us _knowing_ cases where we need to pass data through middle-men LB policies without their knowledge.

> Alright, what about we just plumb the attributes opaquely, but we would build the ResolvedAddresses by explicitly putting in the attributes.

I'm fine with that. It's a bit of a toss-up which approach to go with. The two approaches are clearly not "equivalent" but deciding which is better is more an exercise of predicting the future (which fields will be added).

Reply:
> I'm fine with that.

Actually, I'm not. If this was new code, sure. But changing this when changing everything else seems inappropriate.

Reply:
Alright, alright. I surrender. Will just keep the old way.